firebase-firestore/src/androidTest/java/com/google/firebase/firestore/testutil/CompositeIndexTestHelper.java

```diff
@@ -0,0 +1,178 @@
+    updatedDoc.put(
+        TTL_FIELD,
+        new Timestamp( // Expire test data after 24 hours
+            Timestamp.now().getSeconds() + 24 * 60 * 60, Timestamp.now().getNanoseconds()));
+    return updatedDoc;
+  }
+
+  // Remove test-specific fields from a document.
+  private Map<String, Object> removeTestSpecificFieldsFromDoc(Map<String, Object> doc) {
+    doc.remove(TTL_FIELD);
+    doc.remove(TEST_ID_FIELD);
+    return doc;
+  }
+
+  // Helper method to hash document keys and add test-specific fields for the provided documents.
+  private Map<String, Map<String, Object>> prepareTestDocuments(
+      Map<String, Map<String, Object>> docs) {
+    Map<String, Map<String, Object>> result = new HashMap<>();
+    for (String key : docs.keySet()) {
+      Map<String, Object> doc = addTestSpecificFieldsToDoc(docs.get(key));
+      result.put(toHashedId(key), doc);
+    }
+    return result;
+  }
+
+  // Asserts that the result of running the query while online (against the backend/emulator) is
+  // the same as running it while offline. The expected document Ids are hashed to match the
+  // actual document IDs created by the test helper.
+  @NonNull
+  public void assertOnlineAndOfflineResultsMatch(
+      @NonNull Query query, @NonNull String... expectedDocs) {
+    checkOnlineAndOfflineResultsMatch(query, toHashedIds(expectedDocs));
+  }
+
+  // Adds a filter on test id for a query.
+  @NonNull
+  public Query query(@NonNull Query query_) {
+    return query_.whereEqualTo(TEST_ID_FIELD, testId);
+  }
+
+  // Get document reference from a document key.
+  @NonNull
+  public DocumentReference getDocRef(
+      @NonNull CollectionReference collection, @NonNull String docId) {
+    if (!docId.contains("test-id-")) {
+      docId = toHashedId(docId);
+    }
+    return collection.document(docId);
+  }
+
+  // Adds a document to a Firestore collection with test-specific fields.
+  @NonNull
+  public Task<DocumentReference> addDoc(
+      @NonNull CollectionReference collection, @NonNull Map<String, Object> data) {
+    return collection.add(addTestSpecificFieldsToDoc(data));
+  }
+
+  // Sets a document in Firestore with test-specific fields.
+  @NonNull
+  public Task<Void> setDoc(@NonNull DocumentReference document, @NonNull Map<String, Object> data) {
+    return document.set(addTestSpecificFieldsToDoc(data));
+  }
+
+  @NonNull
+  public Task<Void> updateDoc(
+      @NonNull DocumentReference document, @NonNull Map<String, Object> data) {
+    return document.update(data);
+  }
+
+  @NonNull
+  public Task<Void> deleteDoc(@NonNull DocumentReference document) {
+    return document.delete();
+  }
+
+  // Retrieves a single document from Firestore with test-specific fields removed.
+  @NonNull
+  public Map<String, Object> getDoc(@NonNull DocumentReference document) {
+    DocumentSnapshot docSnapshot = waitFor(document.get());
+    return removeTestSpecificFieldsFromDoc(docSnapshot.getData());
+  }
```

Comment:
I was hoping we could return a DocumentSnapshot here, but that's not easily possible.

We could try doing it using the `MutableDocument.newFoundDocument()`, but then we'd have to use `UserDataReader` and convert the `Map<String, Object>` to `Map<String, Value>` and we'd also have to set `SnapshotVersion.NONE`. All of this to construct a synthetic document snapshot. It's not worth it I think, given that the caller can compare to the expected document's data.

Reply:
Yes, I have wrestled with the types as well, data in a snapshot cannot be directly modified like Web does, neither we can create a new DocumentSnapshot based on the info on hand. Currently, getDoc and getDocs in most of the testings always followed by "...ToValues" or "...ToIds". So focusing on the value itself is good enough for composite index as well.

Reply:
SG